src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -70,37 +69,87 @@ private void refreshCacheItem(final MessageCacheItem cacheItem, Iterator<DataSou
 		}
 	}
 
+	@Deprecated
+	public MessageCacheItem getMessagesOld() {
+		Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
+		return this.getMessages(fallbackLocalesIter);
+	}
+
 	/**
-	 * Get MessageCacheItem from cache.
-	 * The cache is refreshed if MessageCacheItem is expired or not found.
-	 * Pre-configured locale fallback queue is used on failure.
+	 * Get MessageCacheItem from cache. Refreshes the cache in a separate thread if MessageCacheItem is expired or not found.
 	 *
-	 * @return A MessageCacheItem whose message map is one of the items in the following priority-ordered list:
+	 * @return A MessageCacheItem whose message map is the first of the following to be found in the cache:
 	 * <ul>
 	 * 		<li>The messages in the requested locale</li>
-	 * 		<li>The messages in a default locale</li>
+	 * 		<li>The messages in a fallback locale. It iterates over a pre-configured locale fallback priority queue.</li>
 	 * 		<li>The source messages</li>
-	 * 		<li>An empty map</li>
 	 * </ul>
+	 * If none of the above was found in the cache, then it will try to populate the cache and wait for the result.
+	 * It will use the cached data if successful. Otherwise, the MessageCacheItem's message map will be empty.
 	 */
 	public MessageCacheItem getMessages() {
+		MessageCacheItem cacheItem = getMessagesFromCache();
 		Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
-		return this.getMessages(fallbackLocalesIter);
+		if (cacheItem == null) {
+			return this.getMessages(fallbackLocalesIter);
+		} else {
+			this.getMessages(fallbackLocalesIter, false);
+			return cacheItem;
+		}
 	}
 
 	/**
 	 * Get MessageCacheItem from cache.
-	 * The cache is refreshed if MessageCacheItem is expired or not found.
+	 *
+	 * @return A MessageCacheItem whose message map is the first of the following to be found in the cache:
+	 * <ul>
+	 * 		<li>The messages in the requested locale</li>
+	 * 		<li>The messages in a fallback locale. It iterates over a pre-configured locale fallback priority queue.</li>
+	 * 		<li>The source messages</li>
+	 * </ul>
+	 * If none of the above is found in the cache, then tit will return null.
+	 */
+	private MessageCacheItem getMessagesFromCache() {
+		MessageCacheItem cacheItem = null;
+		CacheService cacheService = new CacheService(dto);
+		cacheItem = cacheService.getCacheOfComponent();
+
+		Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
```

Comment:
This isn't a good solution because it will mislead users that the translation of support locale is not ready

Reply:
I think it's wrong behavior. If a locale is supported, its translation should return no matter it has been in cache or not.

Reply:
It will as soon as it is fetched. However, there is this delay that was introduced by adding a VIP service in the picture. This problem has been there even before the new workflow was introduced. Before integrating with VIP, a product application only fetches localized resources from locally bundled files (disk access). When you introduce VIP service, you are automatically introducing a delay (fallback logic, network hops, etc.). So, to mitigate this performance hit, we immediately serve a cached fallback only if the requested locale (supported or not) is not yet cached. 

If cache was initialized at app start, a supported locale would have been pre-cached. This means that the requested locale, not the fallback, will be returned on first request. So this logic in the PR applies mostly to non-supported locales (and to supported locales only if not pre-cached).

If still unclear, let's discuss this offline soon.

Reply:
I can't agree with you. The correctness takes priority over performance.

Reply:
As discussion in slack, please come up with a new solution.

Reply:
What discussion? I am actually asking you for a better solution. :)

Reply:
@xiaochao, regarding your concern about PR 757 — fallback behavior. You have a point.. Let’s talk and try to come up with a better solution.

Hi Jessie, above is what you sent in Slack. Please check. It's better to leave it in GitHub:-)